### PR TITLE
Mapgen road connections improved in the absence of overmap cities

### LINKED ISSUE #48602: Bug: Campgrounds with cars don't spawn with road connections.
# Describe the bug

Campgrounds that have cars dont spawn with road connections.

# Steps To Reproduce

1. Choose wilderness start, campground spawn / Find campground
2. Odds are you will spawn surrounded by trees, miles from any road.
3. Cars that spawn in the campground with a high chance of being drivable cannot be taken out because of tree obstructions and no road path.

# Expected behavior

Since campgrounds spawn with a high percentage of drivable cars, and have facilities on par with state parks, rest stops, etc and a path, they are supposed to be connected to roads. And not surrounded by impassible forest.

This is only for the 2x2 campgrounds which spawn with mostly drivable vehicles and facilities.  The 1x1 campgrounds have no vehicles or facilities, so it makes sense for those to be off grid.

# Screenshots

campgrounds fubar

image

Do not take a screenshot of the crashing window. Please provide the crash logs in the "Additional context" -->

# Versions and configuration

0.E
No Mods

# Additional context

I think this is a result of increased city spacing and size.

### PR DESCRIPTION
**Pull request by stubkan**
#### Summary
Bugfixes "Mapgen road connections improved in the absence of overmap cities"

#### Purpose of change

Adds a function that runs if a city is not found or defined during the road connection phase of placing a special.  It searches up to 50 tiles away for a road to connect to, instead.

#46881
Does not fully eliminate this bug, need another fix to add road connections to mission locations.

Fixes #48602

#### Describe the solution

The function to create a road only runs if a city is defined.  So, in the absence of a city, such as when density and spacing is boosted - no connection roads are made for any special if there is no nearby city.  So, an else function runs and scans up to 50 tiles away for a road to connect to.  Little baby no-name towns/villages also now get connected to roads, which is good.

Things that are more than 50 tiles from a road would still fail to be connected.  Perhaps this could be increased, Is there a reason it shouldn't go too high?

#### Describe alternatives you've considered

I considered changing the place_special function, but this is an easier solution that does not change any pre-existing functions.

#### Testing

I made a few maps, normal spacing maps have the same generation, and this is only really a noticeable bug on larger spacing maps.  On larger spacing maps, stuff that was never connected, is now being connected to roads, such as innawoods campgrounds, military bases, country labs, and so forth.  I have not tested mission generation placements, but will try to soon.

#### Additional context

linking without cities

### CODE REVIEW

File: src/overmap.cpp [hunk 1]
```diff
@@ -4131,12 +4130,20 @@ void overmap::place_special(
...
                     initial_dir = om_direction::add( initial_dir, dir );
                 }
-
+				if ( cit ) {
                 build_connection( cit.pos, rp.xy(), elem.p.z, *elem.connection, must_be_unexplored,
                                   initial_dir );
+	            }
+	// if no city present, search for nearby road within 50 tiles and make connection to it instead
+                else {
+                    for( const tripoint_om_omt &nearby_point : closest_points_first(rp, 50) ) {
```
Comment (Salty-Panda) [inline src/overmap.cpp:4139]:
```suggestion
                    for( const tripoint_om_omt &nearby_point : closest_points_first( rp, 50 ) ) {
```

File: src/overmap.cpp [hunk 2]
```diff
@@ -4131,12 +4130,20 @@ void overmap::place_special(
...
+                        if( check_ot( "road", ot_match_type::contains, nearby_point ) ) {
+                                build_connection(nearby_point.xy(), rp.xy(), elem.p.z, *elem.connection, must_be_unexplored, initial_dir );
```
Comment (Salty-Panda) [inline src/overmap.cpp:4141]:
```suggestion
                                build_connection( nearby_point.xy(), rp.xy(), elem.p.z, *elem.connection, must_be_unexplored, initial_dir );
```

**Comment by actual-nh** [comment]
A modified form of this, in which - at least sometimes - connections were done to roads even if cities were available, might help with #35212 (or at least make it easier to fix).

**Comment by stubkan** [comment]
> A modified form of this, in which - at least sometimes - connections were done to roads even if cities were available, might help with #35212 (or at least make it easier to fix).

There are more things I wish to do with mapgen, definitely will work on that aspect you linked.  Thanks for bringin that one up.  

This does make any location in the country side more guaranteed to be linked to a road, so it is one step toward better countryside mapgen, cos if you happen to get an overworld tile without a city, that tile will have basically nothing connected to the road.

**Comment by stubkan** [comment]
> A modified form of this, in which - at least sometimes - connections were done to roads even if cities were available, might help with #35212 (or at least make it easier to fix).

There are more things I wish to do with mapgen, definitely will work on that aspect you linked.  Thanks for bringin that one up.  

This does make any location in the country side more guaranteed to be linked to a road, so it is one step toward better countryside mapgen, cos if you happen to get an overworld tile without a city, that tile will have basically nothing connected to the road.

**Comment by actual-nh** [comment]
If you put "Fixes #48602", as opposed to just "#48602", it would automatically note this on that issue, and close it once this goes in.

**Pull request merged**